### ISSUE TITLE
Add support for truncated PBKDF2-HMAC-SHA1 hashes

### ISSUE THREAD

**Issue by lokkju** [openwall/john#4991]
In preparation for adding support for extracting and cracking Chrome Sync's custom passwords from the local profile's Nigori.bin file, I'd like to add support for truncated PBKDF2-HMAC-SHA1 hashes. Currently, the format only seems to support 20 byte hash lengths, though the PBKDF2HMAC algorithm supports any length *up to* 20 bytes. Chrome Sync's Nigori keys only use and store 16 bytes of keying material.

As [suggested](https://www.openwall.com/lists/john-users/2022/01/07/2) on the mailing list, I've added support for the pbkdf2-hmac-sha1 format to dynamically determine the key length. This should support keys from 8 (max `ARCH_SIZE`, to optimize `cmp_all`) to 20 bytes, including purposefully short or accidentally truncated hashes.

I've added support for the cpu code paths; I haven't (and probably won't) mess with opencl at the moment.

If the approach sounds good, I'll open a pull request for [pbkdf-hmac-sha1-truncated-hash-support-cpuonly](https://github.com/lokkju/john/tree/pbkdf-hmac-sha1-truncated-hash-support-cpuonly)

**Comment by magnumripper** [comment]
Sounds great to me.

> Currently, the format only seems to support 20 byte hash lengths, though the PBKDF2HMAC algorithm supports any length _up to_ 20 bytes.

PBKDF2 also supports *longer* ones than the base algo's length. If I recall right some/all of our PBKDF2 formats support that, so for SHA-1 it could support eg. 40 bytes but it wouldn't actually calculate more than 20. Then in `cmp_one()`or `cmp_exact()` it would calculate the rest and complain/warn if we got a partial match (because that is so very unlikely). This way we don't lose any speed at all.

Edits: typos

**Comment by magnumripper** [comment]
BTW I believe the current PBKDF2 formats are fairly unified now, so any change to one of them should be very easy to backport to the others. It'd be great if you could do so while at it.

**Comment by lokkju** [comment]
> PBKDF2 also supports _longer_ ones than the base algo's length. If I recall right some/all of our PBKDF2 formats support that, so for SHA-1 it could support eg. 40 bytes but it wouldn't actually calculate more than 20. Then in `cmp_one()`or `cmp_exact()` it would calculate the rest and complain/warn if we got a partial match (because that is so very unlikely). This way we don't lose any speed at all.

Yup, that's what I'm doing: cmp_all compares `ARCH_SIZE`, cmp_one compares keylen (up to max 20), and cmp_exact was already comparing the full thing, and warns if there is a mismatch

> BTW I believe the current PBKDF2 formats are fairly unified now, so any change to one of them should be very easy to backport to the others. It'd be great if you could do so while at it.

There are quite a few different PBKDF2HMAC format plugs, each with their own cmp methods. I'll take a look though.

**Comment by magnumripper** [comment]
BTW this is a dupe of #4492 (for all PBKDF2-HMAC formats) so if you fix all of them, it'd close that issue as well.

**Comment by rbsec** [comment]
389-ds uses PBKDF2-HMAC-SHA256 with a 64 byte salt and a 256 byte key length ([their source code](https://pagure.io/389-ds-base/blob/542287c/f/ldap/servers/plugins/pwdstorage/pbkdf2_pwd.c#_31) and a [python implementation](https://github.com/thesubtlety/389-ds-password-check/blob/master/389ds-pwdcheck.py#L59)) - so it would be definitely be useful to support longer key lengths.

**Comment by solardiz** [comment]
> I've added support for the cpu code paths; I haven't (and probably won't) mess with opencl at the moment.
> 
> If the approach sounds good, I'll open a pull request for [pbkdf-hmac-sha1-truncated-hash-support-cpuonly](https://github.com/lokkju/john/tree/pbkdf-hmac-sha1-truncated-hash-support-cpuonly)

I just took a look. We do want this pull request, please. However, you'll have to mess with the host part of the OpenCL code as well, because you're already changing some host source code shared for CPU and OpenCL formats. You allow the lower lengths, so OpenCL host code's `cmp_*` need to be adapted to support them as well, or people will be getting false negatives (I guess).

Please also take a look at the comment I just added to #4492. It looks like for another PBKDF2 format, we somehow just happened to support limited truncation (16+ bytes)? Need to see what's different and unify our approach and code.
